spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala

```diff
@@ -41,3 +60,160 @@ trait ClusteredTableUtilsBase extends DeltaLogging {
 }
 
 object ClusteredTableUtils extends ClusteredTableUtilsBase
+
+/**
+ * A container for clustering information. Copied from OSS Spark.
+ *
+ * This class will be removed when we integrate with OSS Spark's CLUSTER BY implementation.
+ * @see https://github.com/apache/spark/pull/42577
+ *
+ * @param columnNames the names of the columns used for clustering.
+ */
+case class ClusterBySpec(columnNames: Seq[NamedReference]) {
+  override def toString: String = toJson
+
+  def toJson: String =
+    ClusterBySpec.mapper.writeValueAsString(columnNames.map(_.fieldNames))
+}
+
+object ClusterBySpec {
+  private val mapper = {
+    val ret = new ObjectMapper() with ClassTagExtensions
+    ret.setSerializationInclusion(Include.NON_ABSENT)
+    ret.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    ret.registerModule(DefaultScalaModule)
+    ret
+  }
+
+  // ClassTag is added to avoid the "same type after erasure" issue with the case class.
+  def apply[_: ClassTag](columnNames: Seq[Seq[String]]): ClusterBySpec = {
```

Comment:
Looks like this is only used by parser, so it will be removed when we integrated with OSS? If so, can you add a comment that this will be removed (or just put this functionality inside the caller to make it easy to remember); note that I don't see this function in OSS Spark either.

Reply:
Moved to its own package `clustering.temp` and file `ClusterBySpec.scala`.